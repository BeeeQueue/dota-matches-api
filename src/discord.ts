import { mande } from "mande"
import { encode } from "msgpackr/pack"

import { badRequest, ok } from "@worker-tools/response-creators"

const SCOPES = ["bot", "applications.commands"]
const TOKEN_CACHE_KEY = "discord-token"

const discordClient = mande("https://discord.com/api", {
  responseAs: "json",
  headers: {
    "User-Agent": `dota-matches-api (https://github.com/beeequeue/dota-matches-api, ${GIT_SHA})`,
  },
})

const fetchToken = async (): Promise<{ access_token: string; expires_in: number }> => {
  const response = await fetch("https://discord.com/api/v10/oauth2/token", {
    method: "POST",
    headers: {
      "User-Agent": `dota-matches-api (https://github.com/beeequeue/dota-matches-api, ${GIT_SHA})`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams({
      grant_type: "client_credentials",
      client_id: DISCORD_CLIENT_ID,
      client_secret: DISCORD_CLIENT_SECRET,
      scopes: SCOPES.join(" "),
    }),
  })

  if (!response.ok) {
    console.error(response.body)
    throw response
  }

  return await response.json()
}

const getToken = async (env: Env): Promise<string> => {
  const cached = await env.CACHE.get(TOKEN_CACHE_KEY)
  if (cached != null) {
    return cached
  }

  const { access_token, expires_in } = await fetchToken()

  await env.CACHE.put(TOKEN_CACHE_KEY, access_token, {
    expirationTtl: expires_in,
  })

  return access_token
}

export type Guild = {
  id: string
  vanityUrlCode?: unknown
  subscriptions: Array<{ channelId: string; teams: string[] }>
}

type RegisterGuildOptions = {
  code: string
  guildId: string
  permissions: string
}

type TokenResponse = {
  access_token: string
  refresh_token: string
  expires_in: number
  guild: {
    id: string
    vanityUrlCode?: unknown
  }
}

const registerGuild = async (
  env: Env,
  { code, guildId, permissions }: RegisterGuildOptions,
): Promise<Response> => {
  const response = await fetch("https://discord.com/api/v10/oauth2/token", {
    method: "POST",
    headers: {
      "User-Agent": `dota-matches-api (https://github.com/beeequeue/dota-matches-api, ${GIT_SHA})`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      client_id: DISCORD_CLIENT_ID,
      client_secret: DISCORD_CLIENT_SECRET,
      code,
    }),
  })

  if (!response.ok) {
    console.error(response.body)
    throw response
  }

  const { access_token, expires_in, guild }: TokenResponse = await response.json()

  if (permissions !== "34880") {
    console.log(`Got invalid permissions: ${permissions}\n${access_token}`)
    await Discord.leaveGuild(access_token, guildId)

    return badRequest("All the required permissions we're not given.")
  }

  const guildRegistration: Guild = {
    id: guild.id,
    vanityUrlCode: guild.vanityUrlCode,
    subscriptions: [],
  }
  await env.WEBHOOKS.put(guildId, encode(guildRegistration) as ArrayBuffer)
  await env.CACHE.put(guildId, access_token, {
    expirationTtl: expires_in,
  })

  return ok()
}

const getAuthorizeUrl = () => {
  /* eslint-disable @typescript-eslint/naming-convention */
  const searchParams = new URLSearchParams({
    response_type: "code",
    client_id: DISCORD_CLIENT_ID,
    scope: SCOPES.join(" "),
    permissions: "34880",
  })

  return new URL(`https://discord.com/api/oauth2/authorize?${searchParams.toString()}`)
}

const leaveGuild = (token: string, guildId: string) => {
  console.log(`Leaving guild ${guildId}`)
  return discordClient.delete(`/v10/users/@me/guilds@${guildId}`, {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  })
}

export const Discord = { getAuthorizeUrl, registerGuild, leaveGuild }
