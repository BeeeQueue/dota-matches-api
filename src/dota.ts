import { mande } from "mande"
import { HTMLElement, parse } from "node-html-parser"
import PQueue from "p-queue"

const THREE_HOURS = 60 * 60 * 3

type Team = {
  name: string | null
  url: string | null
}

export type Match = {
  teams: [Team, Team]
  matchType: string | null
  startsAt: Date | null
  streamUrl: string | null
}

const liquipediaQueue = new PQueue({ intervalCap: 1, interval: 30_000 })
const liquipediaClient = mande("https://liquipedia.net/dota2", {
  responseAs: "json",
  query: {
    format: "json",
  },
  headers: {
    /* eslint-disable @typescript-eslint/naming-convention */
    "Accept-Encoding": "gzip",
  },
})

type LiquipediaBody = {
  parse: {
    title: string
    pageid: number
    revid: number
    text: {
      ["*"]: string
    }
    langlinks: string[]
    categories: unknown[]
    links: unknown[]
    templates: unknown[]
    images: string[]
    externallinks: string[]
  }
}

const extractTeam = (team$: HTMLElement): Team => {
  const name = team$
    .querySelector(".team-template-text > a")
    ?.attrs?.title?.replace(/\(.*?\)/g, "")
    ?.trim()
  const urlPath = team$.querySelector("[href^='/dota2/']")?.attrs?.href

  return {
    name: name ?? null,
    url: urlPath ? `https://liquipedia.com${urlPath}` : null,
  }
}

const fetchTeamsData = async (country: string) => {
  const data = await liquipediaQueue.add(() =>
    liquipediaClient.get<LiquipediaBody>("/api.php", {
      headers: {
        "User-Agent": `dota-matches-api-${country}/${GIT_SHA}`,
      },
      query: {
        action: "parse",
        page: "Liquipedia:Upcoming_and_ongoing_matches",
      },
    }),
  )

  const root = parse(data.parse.text["*"])

  const $matches = root.querySelectorAll("[data-toggle-area-content='2'] > table")
  if ($matches.length === 0) return []

  const matches = $matches.map<Match>(($match) => {
    const teamLeft$ = $match.querySelector(".team-left")!
    const teamRight$ = $match.querySelector(".team-right")!
    const versus$ = $match.querySelector(".versus")!
    const meta$ = $match.querySelector(".timer-object")!

    const matchType = versus$.querySelector("abbr")?.textContent
    const startTime = meta$.attrs["data-timestamp"]
    const streamName = meta$.attrs["data-stream-twitch"]

    return {
      teams: [extractTeam(teamLeft$), extractTeam(teamRight$)],
      matchType: matchType ?? null,
      startsAt: startTime ? new Date(Number(startTime) * 1000) : null,
      streamUrl: streamName ? `https://www.twitch.tv/${streamName}` : null,
    }
  })

  return matches
}

const CACHE_KEY = "liquipedia-matches"

const getMatches = async (env: Env, country: string): Promise<Match[]> => {
  console.log(`Getting match data...`)

  const cached = await env.CACHE.get(CACHE_KEY)
  if (cached != null) {
    return JSON.parse(cached)
  }

  const matches = await fetchTeamsData(country)

  await env.CACHE.put(CACHE_KEY, JSON.stringify(matches), {
    expirationTtl: THREE_HOURS,
  })

  return matches
}

export const Dota = { getMatches }
