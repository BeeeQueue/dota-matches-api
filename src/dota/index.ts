import { differenceInHours } from "date-fns"
import { mande, MandeError } from "mande"

import { isNotNil } from "../utils"

import { teamsQuery } from "./queries"
import { TeamsQuery, TeamsQueryVariables } from "./types"

export type Match = {
  id: number
  startsAt: Date
  teams: [
    {
      id: number
      name: string
    },
    {
      id: number
      name: string
    },
  ]
}

const isMatchRelevant = (match: Match): boolean =>
  Math.abs(differenceInHours(new Date(), match.startsAt)) <= 24

const cleanMatchData = (
  series: NonNullable<
    NonNullable<NonNullable<TeamsQuery["teams"]>[number]>["series"]
  >[number],
): Match | null => {
  if (series == null) return null

  const firstGameStartTime = series.matches?.reduce(
    (accum, match) =>
      (accum ?? Number.NEGATIVE_INFINITY) <
      (match!.startDateTime ?? Number.POSITIVE_INFINITY)
        ? match!.startDateTime
        : accum,
    null as number | null,
  )

  if (firstGameStartTime == null) return null

  return {
    id: series.id,
    startsAt: new Date(firstGameStartTime * 1000),
    teams: [
      {
        id: series.teamOne!.id,
        name: series.teamOne!.name!,
      },
      {
        id: series.teamTwo!.id,
        name: series.teamTwo!.name!,
      },
    ],
  }
}

const stratzClient = mande("https://api.stratz.com", {
  responseAs: "json",
  headers: {
    Authorization: `Bearer ${STRATZ_TOKEN}`,
  },
})

const fetchLeagueMatches = async (ids: number[]): Promise<TeamsQuery> => {
  if (ids.length === 0) {
    return {
      teams: [],
    }
  }

  console.log(`Fetching ${ids.join(", ")}...`)

  const variables: TeamsQueryVariables = {
    ids,
  }
  const result = await stratzClient
    .post<{ data: TeamsQuery }>("/graphql", {
      query: teamsQuery,
      variables,
    })
    .catch((error: MandeError) => error)

  if (result instanceof Error) {
    const body = await result.response
      .json<Record<string, unknown>>()
      .catch(() => result.response.text().catch(() => null))
    console.error({
      status: result.response.status,
      body,
    })
    throw new Error(
      `Failed to fetch matches: ${result.response.status}, ${JSON.stringify(body!)}`,
    )
  }

  return result.data
}

const getTeamsData = async (env: Env, ids: number[]) => {
  console.log(`Getting match data for ${ids.join(", ")}...`)

  const matches: Match[] = []
  const notCachedIds = new Set<number>()

  await Promise.all(
    ids.map(async (id) => {
      const cached = await env.CACHE.get(id.toString())
      if (cached != null) {
        matches.push(...(JSON.parse(cached) as Match[]))
        return
      }

      notCachedIds.add(id)
    }),
  )
  console.log(`Found ${ids.length - notCachedIds.size} cached teams`)

  if (notCachedIds.size > 0) {
    console.log(`Fetching ${notCachedIds.size} matches...`)

    const data = await fetchLeagueMatches([...notCachedIds])
    const matchesByTeam = (data.teams ?? []).reduce((accum, team) => {
      if (team == null || team.series == null) return accum

      accum[team.id.toString()] ??= []

      const newMatches = team.series
        .map(cleanMatchData)
        .filter((match): match is Match => isNotNil(match) && isMatchRelevant(match))
      accum[team.id.toString()].push(...newMatches)

      return accum
    }, {} as Record<string, Match[]>)

    console.log(`Caching ${Object.keys(matchesByTeam).length} teams...`)
    await Promise.all(
      Object.entries(matchesByTeam).map(async ([teamId, newMatches]) => {
        await env.CACHE.put(teamId, JSON.stringify(newMatches), {
          expirationTtl: 60 * 60 * 3,
        })

        matches.push(...newMatches)
      }),
    )
  }

  console.log(`Deduping and returning...`)
  const matchIds = new Set<number>()
  return matches.filter((match) => {
    if (!matchIds.has(match.id)) {
      matchIds.add(match.id)
      return true
    }

    return false
  })
}

export const Dota = { getTeamsData }
