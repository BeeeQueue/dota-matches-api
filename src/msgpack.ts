/* eslint-disable import/no-duplicates */
import { pack } from "msgpackr/pack"
import { unpack } from "msgpackr/unpack"

export const encode = <T = unknown>(data: T) => pack(data) as ArrayBuffer

export const decode = async <T>(data: Response | R2ObjectBody | Uint8Array) => {
  if (typeof (data as Response).arrayBuffer === "function") {
    data = new Uint8Array(await (data as Response).arrayBuffer())
  }

  return unpack(data as Uint8Array) as T
}
