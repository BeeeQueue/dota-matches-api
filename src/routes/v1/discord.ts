import { InteractionResponseType, InteractionType, verifyKey } from "discord-interactions"
import { IHTTPMethods, Router } from "itty-router"

import { badRequest, ok, temporaryRedirect } from "@worker-tools/response-creators"

import { Discord } from "../../discord"
import {
  handleFollowCommand,
  handleListCommand,
  InteractionsBody,
} from "../../discord/commands"
import { json } from "../../utils"

export const discordRouter = Router<Request, IHTTPMethods>({ base: "/v1/discord" })

discordRouter.get("/", () => temporaryRedirect(Discord.getAuthorizeUrl()))

discordRouter.post("/interactions", async (request: Request, env: Env) => {
  const body = await request.text()
  const signature = request.headers.get("x-signature-ed25519")
  const timestamp = request.headers.get("x-signature-timestamp")
  if (!verifyKey(body, signature!, timestamp!, DISCORD_PUBLIC_KEY)) {
    return badRequest("Invalid request signature")
  }

  const parsedBody = JSON.parse(body) as InteractionsBody
  const { type, data } = parsedBody

  if (type === InteractionType.PING) {
    return json({ type: InteractionResponseType.PONG })
  }

  if (type === InteractionType.APPLICATION_COMMAND && data != null) {
    switch (data.name) {
      case "follow":
        return handleFollowCommand(env, parsedBody)

      case "unfollow":
        return ok()

      case "follows":
        return handleListCommand(env, parsedBody)
    }

    return badRequest("Invalid command")
  }

  return ok()
})

const isValidCallback = (params: URLSearchParams) =>
  params.has("code") && params.has("guild_id") && params.has("permissions")

discordRouter.get("/callback", async (request, env: Env) => {
  const url = new URL(request.url)

  if (!isValidCallback(url.searchParams)) {
    return badRequest("Missing or invalid callback parameters")
  }

  return Discord.registerGuild(env, {
    code: url.searchParams.get("code")!,
    guildId: url.searchParams.get("guild_id")!,
    permissions: url.searchParams.get("permissions")!,
  })
})
