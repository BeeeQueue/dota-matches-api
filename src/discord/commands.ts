import {
  APIChatInputApplicationCommandInteraction,
  APIInteractionResponse,
  ApplicationCommandOptionType,
  InteractionResponseType,
} from "discord-api-types/v10"

import { badRequest } from "@worker-tools/response-creators"

import { decode, encode } from "../msgpack"
import { json } from "../utils"

import { Guild } from "./index"

export enum Command {
  Follow = "follow",
  Unfollow = "unfollow",
  List = "follows",
}

// enum MessageComponentId {}

export const handleFollowCommand = async (
  env: Env,
  body: APIChatInputApplicationCommandInteraction,
) => {
  const guildId = body.guild_id!
  const teamNames = body.data
    .options!.map((option) =>
      option.type === ApplicationCommandOptionType.String ? option.value : null,
    )
    .filter((value): value is string => value != null)

  const guildObject = await env.WEBHOOKS.get(guildId)
  if (guildObject == null) {
    return badRequest("Guild not registered.")
  }

  const guild = await decode<Guild>(guildObject)
  guild.subscriptions[body.channel_id] ??= []

  for (const subscribedChannelId in guild.subscriptions) {
    if (subscribedChannelId === body.channel_id) {
      guild.subscriptions[body.channel_id] = [
        // New teams are added _first_ in the array, to make deduping easier
        ...new Set([...teamNames, ...guild.subscriptions[body.channel_id]]),
      ]
    }
  }

  await env.WEBHOOKS.put(guildId, encode(guild))

  const response: APIInteractionResponse = {
    type: InteractionResponseType.ChannelMessageWithSource,
    data: {
      content: `Okay, I will now notify you those teams' matches.`,
    },
  }

  return json(response)
}

export const handleListCommand = async (
  env: Env,
  body: APIChatInputApplicationCommandInteraction,
) => {
  const guildObject = await env.WEBHOOKS.get(body.guild_id!)
  if (guildObject == null) {
    return badRequest("Guild not registered.")
  }

  const guild = await decode<Guild>(guildObject)
  if (guild.subscriptions[body.channel_id] == null) {
    const response: APIInteractionResponse = {
      type: InteractionResponseType.ChannelMessageWithSource,
      data: {
        content: `This channel is not following any teams. Follow some with '/follow <team>'!`,
      },
    }

    return json(response)
  }

  const response: APIInteractionResponse = {
    type: InteractionResponseType.ChannelMessageWithSource,
    data: {
      content: `This channel is following:
\`\`\`
${guild.subscriptions[body.channel_id].join("\n")}
\`\`\``,
    },
  }

  return json(response)
}
