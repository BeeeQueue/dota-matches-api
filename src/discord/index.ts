import { format } from "date-fns"
import {
  RESTPostAPIChannelMessageJSONBody,
  RESTPostAPIChannelMessageResult,
  RESTPostAPIChannelThreadsJSONBody,
  RESTPostAPIChannelThreadsResult,
  RESTPostOAuth2AccessTokenURLEncodedData,
  RESTPostOAuth2AccessTokenWithBotAndGuildsScopeResult,
  Routes,
  ThreadAutoArchiveDuration,
} from "discord-api-types/v10"
import { mande } from "mande"

import { badRequest, ok } from "@worker-tools/response-creators"

import { encode } from "../msgpack"

const SCOPES = ["bot", "applications.commands"]
const BOT_PERMISSIONS = "309237647360"
const TOKEN_CACHE_KEY = "discord-token"

const baseUrl = "https://discord.com/api/v10"
const userAgent = `dota-matches-api (https://github.com/beeequeue/dota-matches-api, ${GIT_SHA})`
const discordClient = mande(baseUrl, {
  responseAs: "json",
  headers: {
    "User-Agent": userAgent,
  },
})

export type Guild = {
  id: string
  vanityUrlCode?: unknown
  /** New teams are added _first_ in the array, to make deduping easier later */
  subscriptions: Record<string, string[]> // channelId: teamName[]
}

type RegisterGuildOptions = {
  code: string
  guildId: string
  permissions: string
}

const registerGuild =
  (env: Env) =>
  async ({ code, guildId, permissions }: RegisterGuildOptions): Promise<Response> => {
    const urlEncodedBody: RESTPostOAuth2AccessTokenURLEncodedData = {
      grant_type: "authorization_code",
      client_id: env.DISCORD_CLIENT_ID,
      client_secret: env.DISCORD_CLIENT_SECRET,
      code,
    }
    const response = await fetch(`${baseUrl}${Routes.oauth2TokenExchange()}`, {
      method: "POST",
      headers: {
        "User-Agent": userAgent,
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams(urlEncodedBody as Record<string, any>),
    })

    if (!response.ok) {
      console.error({
        status: response.status,
        body: await response.json(),
      })
      throw response
    }

    const body: RESTPostOAuth2AccessTokenWithBotAndGuildsScopeResult =
      await response.json()
    const { access_token, expires_in, guild } = body

    if (permissions !== BOT_PERMISSIONS) {
      console.log(`Got invalid permissions: ${permissions}`)
      await leaveGuild(env, guildId)

      return badRequest("All the required permissions were not given.")
    }

    const guildRegistration: Guild = {
      id: guild.id,
      subscriptions: {},
      vanityUrlCode: guild.vanity_url_code ?? null,
    }
    await env.WEBHOOKS.put(guildId, encode(guildRegistration))
    await env.CACHE.put(TOKEN_CACHE_KEY, access_token, {
      expirationTtl: expires_in,
    })

    return ok()
  }

const getAuthorizeUrl = (env: Env) => () => {
  /* eslint-disable @typescript-eslint/naming-convention */
  const searchParams = new URLSearchParams({
    response_type: "code",
    client_id: env.DISCORD_CLIENT_ID,
    scope: SCOPES.join(" "),
    permissions: BOT_PERMISSIONS,
  })

  return new URL(`${baseUrl}${Routes.oauth2Authorization()}?${searchParams.toString()}`)
}

const leaveGuild = async (env: Env, guildId: string) => {
  console.log(`Leaving guild ${guildId}`)
  return discordClient.delete(Routes.userGuild(guildId), {
    headers: {
      Authorization: `Bot ${env.DISCORD_BOT_TOKEN}`,
    },
  })
}

const createThread = (env: Env) => async (channelId: string) => {
  console.log(`Sending message to ${channelId}`)

  const body: RESTPostAPIChannelThreadsJSONBody = {
    type: 11, // public thread
    name: `Match Schedule ${format(new Date(), "MMM ddd")}`,
    auto_archive_duration: ThreadAutoArchiveDuration.OneDay,
  }
  return discordClient.post<RESTPostAPIChannelThreadsResult>(Routes.threads(channelId), {
    body,
    headers: {
      Authorization: `Bot ${env.DISCORD_BOT_TOKEN}`,
    },
  })
}

const sendMessage = (env: Env) => async (parentId: string, content: string) => {
  console.log(`Sending message to ${parentId}`)

  const body: RESTPostAPIChannelMessageJSONBody = {
    content,
  }
  return discordClient.post<RESTPostAPIChannelMessageResult>(
    Routes.channelMessages(parentId),
    {
      body,
      headers: {
        Authorization: `Bot ${env.DISCORD_BOT_TOKEN}`,
      },
    },
  )
}

export const createDiscordClient = (env: Env) => ({
  getAuthorizeUrl: getAuthorizeUrl(env),
  registerGuild: registerGuild(env),
  createThread: createThread(env),
  sendMessage: sendMessage(env),
  leaveGuild,
})
