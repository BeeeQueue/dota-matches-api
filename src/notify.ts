import { addDays, format, isBefore } from "date-fns"
import { APIEmbed, APIEmbedField } from "discord-api-types/v10"
import { indexBy } from "remeda"

import { createDiscordClient, Guild } from "./discord"
import { Dota, Match } from "./dota"
import { decode } from "./msgpack"

const orEmpty = <T>(check: T | null | undefined, value: string) => (check ? value : "")

export const formatMatchToEmbedField = (match: Match): APIEmbedField => {
  const teams = match.teams.map((team) => `**${team!.name!}**`).join(" _vs_ ")

  const startsAtUnix = match.startsAt
    ? Math.round(new Date(match.startsAt).getTime() / 1000)
    : null
  const matchType = orEmpty(match.matchType, `${match.matchType!} `)
  const startsAt = orEmpty(
    startsAtUnix,
    `@<t:${startsAtUnix!}:t> (<t:${startsAtUnix!}:R>)`,
  )
  const leagueName = orEmpty(match.leagueName, `${match.leagueName!}`)
  const separator = match.leagueName != null && match.streamUrl != null ? " | " : ""
  const streamLink = orEmpty(match.streamUrl, `[Stream](${match.streamUrl!})`)

  return {
    name: teams,
    value: `${matchType}${startsAt}\n${leagueName}${separator}${streamLink}`,
  }
}

export const notifier: ExportedHandlerScheduledHandler<Env> = async (
  _controller,
  env: Env,
) => {
  const unfilteredMatches = await Dota.getMatches(env, "main")
  const relevantMatches = unfilteredMatches.filter(
    (match) =>
      match.startsAt != null &&
      match.teams[0]?.name != null &&
      match.teams[1]?.name != null &&
      isBefore(new Date(match.startsAt), addDays(Date.now(), 1)),
  )
  const matches = indexBy(relevantMatches, (match) => match.hash)
  const matchesByTeam = relevantMatches.reduce((accum, match) => {
    if (match.teams[0]?.name != null) {
      accum[match.teams[0].name] ??= new Set()
      accum[match.teams[0].name].add(match.hash)
    }
    if (match.teams[1]?.name != null) {
      accum[match.teams[1].name] ??= new Set()
      accum[match.teams[1].name].add(match.hash)
    }

    return accum
  }, {} as Record<string, Set<string>>)

  const guildObjects = await env.WEBHOOKS.list()

  const guilds = await Promise.all(
    guildObjects.objects.map(async ({ key }) => {
      const obj = (await env.WEBHOOKS.get(key))!
      return decode<Guild>(obj)
    }),
  )

  const channelsMatches: Array<[string, Set<string>]> = []
  for (const guild of guilds) {
    for (const channelId in guild.subscriptions) {
      const followedTeams = Object.values(guild.subscriptions[channelId])
      const matchesToAdd = new Set(
        followedTeams.reduce(
          (accum, team) => [...accum, ...(matchesByTeam[team] ?? [])],
          [] as string[],
        ),
      )

      if (matchesToAdd.size > 0) {
        channelsMatches.push([channelId, matchesToAdd])
      }
    }
  }

  const cachedFields = new Map<string, APIEmbedField>()
  const getFieldsForChannel = (channelId: string, matchHashes: Set<string>) => {
    const fields = [...matchHashes].map((hash) => {
      if (cachedFields.has(channelId)) {
        return cachedFields.get(channelId)!
      }

      const matchField = formatMatchToEmbedField(matches[hash])
      cachedFields.set(hash, matchField)
      return matchField
    })

    return fields
  }

  const embedTemplate: Omit<APIEmbed, "fields"> = {
    title: `Today's matches - ${format(new Date(), "MMM do")}`,
    url: "https://liquipedia.net/dota2/Liquipedia:Upcoming_and_ongoing_matches",
    thumbnail: {
      url: "https://f003.backblazeb2.com/file/bq-files/2022/07/point.webp",
    },
    footer: {
      text: "Data provided by Liquipedia",
      icon_url:
        "https://liquipedia.net/commons/extensions/TeamLiquidIntegration/resources/pagelogo/liquipedia_icon_menu.png",
    },
  }

  const messages = channelsMatches.map(([channelId, matchHashes]) => {
    // const matchLines = getMessageLinesForChannel(channelId, matchHashes)

    const embed: APIEmbed = {
      ...embedTemplate,
      fields: getFieldsForChannel(channelId, matchHashes),
    }

    return [channelId, embed] as const
  })

  const discordClient = createDiscordClient(env)
  await Promise.all(
    messages.map(async ([channelId, embed]) => {
      // const { id: threadId } = await discordClient.createThread(channelId)

      await discordClient.sendMessage(channelId, embed)
    }),
  )
}
