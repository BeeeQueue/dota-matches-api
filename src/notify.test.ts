import { addDays } from "date-fns"
import { MockAgent, setGlobalDispatcher } from "undici"
import { beforeAll, beforeEach, describe, expect, it, vi } from "vitest"

import { Guild } from "./discord"
import { CACHE_KEY, Match } from "./dota"
import matchesFixture from "./fixtures/matches.json"
import { encode } from "./msgpack"
import { formatMatchToMessageLine, notifier } from "./notify"

const GUILD_ID = "987613986523"
const CHANNEL_ID = "0986526095326812"

let agent = new MockAgent()

beforeEach(() => {
  agent = new MockAgent()
  agent.disableNetConnect()
  setGlobalDispatcher(agent)
})

describe.only("formatMatchToMessageLine", () => {
  beforeAll(() => {
    vi.setSystemTime(new Date("2022-07-31T14:00:00.000Z"))

    return () => vi.setSystemTime(vi.getRealSystemTime())
  })

  it.each(matchesFixture)("should format matches correctly %#", (match) => {
    const message = formatMatchToMessageLine(match as Match)

    expect(message).toMatchSnapshot()
  })
})

describe("notifier", () => {
  beforeEach(() => {
    const discordAgent = agent.get("https://discord.com")

    discordAgent.intercept({ path: "/api/v10/oauth2/token", method: "POST" }).reply(200, {
      access_token: "access_token",
      expires_in: 60,
    })

    discordAgent
      .intercept({ path: `/api/v10/channels/${CHANNEL_ID}/messages`, method: "POST" })
      .reply(200, {
        idunno: false,
      })
  })

  it.only("test", async () => {
    const env = (await miniflare.getBindings()) as Env

    const now = new Date()
    const matches: Match[] = [
      {
        hash: "1",
        matchType: "Bo2",
        teams: [
          { name: "Team Liquid", url: "url" },
          { name: "Nigma Galaxy", url: "url" },
        ],
        startsAt: addDays(now, 0.25).toISOString(),
        leagueName: null,
        streamUrl: null,
      },
      {
        hash: "2",
        matchType: "Bo2",
        teams: [
          { name: "Nigma Galaxy", url: "url" },
          { name: "OG", url: "url" },
        ],
        startsAt: addDays(now, 0.5).toISOString(),
        leagueName: null,
        streamUrl: null,
      },
      {
        hash: "3",
        matchType: "Bo2",
        teams: [
          { name: "something", url: "url" },
          { name: "someone", url: "url" },
        ],
        startsAt: addDays(now, 0.35).toISOString(),
        leagueName: null,
        streamUrl: null,
      },
      {
        hash: "4",
        matchType: "Bo2",
        teams: [
          { name: "OG", url: "url" },
          { name: "someone", url: "url" },
        ],
        startsAt: addDays(now, 4).toISOString(),
        leagueName: null,
        streamUrl: null,
      },
    ]
    await env.CACHE.put(CACHE_KEY, JSON.stringify(matches))

    await env.WEBHOOKS.put(
      GUILD_ID,
      encode<Guild>({
        id: GUILD_ID,
        subscriptions: {
          [CHANNEL_ID]: ["Team Liquid", "OG"],
        },
      }),
    )

    const result = await notifier(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any,@typescript-eslint/no-unsafe-argument
      {} as any,
      env,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any,@typescript-eslint/no-unsafe-argument
      {} as any,
    )

    expect(result).toBeUndefined()
  })
})
