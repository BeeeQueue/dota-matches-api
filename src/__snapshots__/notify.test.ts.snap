// Vitest Snapshot v1

exports[`formatMatchToEmbedField > should format matches correctly 0 1`] = `
{
  "name": "**Gaimin Gladiators** _vs_ **Brame**",
  "value": "Bo3 @<t:1659279600:t> (<t:1659279600:R>)
Dota 2 Champions League Season 13",
}
`;

exports[`formatMatchToEmbedField > should format matches correctly 1 1`] = `
{
  "name": "**TSM** _vs_ **Quincy Crew**",
  "value": "Bo3 @<t:1659303000:t> (<t:1659303000:R>)
ESL One Malaysia 2022 North America: Closed Qualifier",
}
`;

exports[`formatMatchToEmbedField > should format matches correctly 2 1`] = `
{
  "name": "**Fnatic** _vs_ **OG**",
  "value": "Bo2 @<t:1659625200:t> (<t:1659625200:R>)
PGL Arlington Major 2022 | [Stream](https://www.twitch.tv/PGL_Dota2_EN3)",
}
`;
